Runtime/LLMChatHistory.cs

```diff
@@ -0,0 +1,139 @@
+                await Load();
+            }
+            else {
+                _chatHistory = new List<ChatMessage>();
+            }
+        }
+
+        /// <summary>
+        /// Appends a new message to the end of this chat.
+        /// </summary>
+        public async Task AddMessage(string role, string content)
+        {
+            await WithChatLock(async () => {
+                await Task.Run(() => _chatHistory.Add(new ChatMessage { role = role, content = content }));
+            });
+
+            // Save our newly updated chat history to the file system
+            _ = Save();
+        }
+
+        public List<ChatMessage> GetChatMessages() {
+            return new List<ChatMessage>(_chatHistory);
+        }
+
+        /// <summary>
+        /// Saves the chat history to the file system
+        /// </summary>
+        public async Task Save()
+        {
+            // If no filename has been provided, create one
+            if (ChatHistoryFilename == string.Empty) {
+                ChatHistoryFilename = $"chat_{Guid.NewGuid()}";
+            }
+
+            string filePath = GetChatHistoryFilePath();
+            string directoryName = Path.GetDirectoryName(filePath);
+
+            // Ensure the directory exists
+            if (!Directory.Exists(directoryName)) Directory.CreateDirectory(directoryName);
+
+            // Save the chat history as json
+            await WithChatLock(async () => {
+                string json = JsonUtility.ToJson(new ChatListWrapper { chat = _chatHistory });
+                await File.WriteAllTextAsync(filePath, json);
+            });
+        }
+
+        /// <summary>
+        /// Load the chat history from the file system
+        /// </summary>
+        public async Task Load()
+        {
+            string filePath = GetChatHistoryFilePath();
+
+            if (!File.Exists(filePath))
+            {
+                LLMUnitySetup.LogError($"File {filePath} does not exist.");
+                return;
+            }
+
+            // Load the chat from the json file
+            await WithChatLock(async () => {
+                string json = await File.ReadAllTextAsync(filePath);
+                _chatHistory = JsonUtility.FromJson<ChatListWrapper>(json).chat;
+                LLMUnitySetup.Log($"Loaded {filePath}");
+            });
+        }
+
+        /// <summary>
+        /// Clears out the current chat history.
+        /// </summary>
+        public async Task Clear() {
+            await WithChatLock(async () => {
+                await Task.Run(() => _chatHistory.Clear());
+            });
+
+            _ = Save();
+        }
+
+        public bool IsEmpty() {
+            return _chatHistory?.Count == 0;
+        }
+
+        protected string GetChatHistoryFilePath()
+        {
+            return Path.Combine(Application.persistentDataPath, ChatHistoryFilename + ".json").Replace('\\', '/');
+        }
+
+        protected async Task WithChatLock(Func<Task> action) {
```

Comment:
Nice function! It could be implemented I think with simple lock, without needing the async / await functionality, something like:
```c#
        readonly object chatLock = new object();

        protected async Task WithChatLock(Func<Task> action) {
        {
            lock (chatLock)
            {
                action();
            }
        }
```

Reply:
Because the actions are async (and we don't want to block on them), I would think we couldn't use the C# "lock" functionality: https://stackoverflow.com/questions/20084695/lock-and-async-method-in-c-sharp

Reply:
The LLMChatHistory doesn't need to have async actions.
It is easier to handle with simple lock and people don't need to await on functions making it easier to integrate in their code.
I tested it and it works file with lock: https://pastebin.com/3pjvr4ge

Reply:
The LLMChatHistory doesn't need to have async actions.
It is easier to handle with simple lock and people don't need to await on functions making it easier to integrate in their code.
I tested it and it works fine with lock: https://pastebin.com/3pjvr4ge